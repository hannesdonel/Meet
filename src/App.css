@import url('https://fonts.googleapis.com/css2?family=Prompt:ital,wght@0,300;0,500;0,700;1,300;1,500;1,700&display=swap');

:root {
  --font-size-smaller: 14px;
  --font-size-standard: 16px;
  --font-size-bigger: 18px;
  --border-radius: 5px;
  --border-color: rgb(204, 204, 204);
  --light-shadow: 0 1px 5px rgba(0, 0, 0, 0.05);
  --heavy-shadow: 0 1px 15px rgba(0, 0, 0, 0.1);
  --light-grey: rgba(238, 238, 238, 0.5);
  --light-grey-no-opacity: rgb(238, 238, 238);
}

* {
  font-family: 'Prompt', sans-serif;
  font-size: var(--font-size-standard);
  box-sizing: border-box;
}

.App {
  display: flex;
  flex-direction: column;
  text-align: center;
  padding: 10px 5% 2%;
}

ul {
  padding: 0;
  list-style: none;
}

p {
  margin: 5px 0;
}

.display-none {
  display: none;
}

.loader-container {
  display: flex;
  align-items: center;
  justify-content: center;
  min-height: 100vh;
}

.showSuggestions {
  display: block;
}

input[type="text"],
input[type="number"] {
  padding: 8px 6px 8px 10px;
  border-radius: var(--border-radius);
  outline: none;
  border: 1px solid #ccc;
  line-height: 20px;
  align-self: center;
  height: 38px;
}

#search-bar-container {
  position: -webkit-fixed;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  width: 100%;
  padding: 0 5%;
  z-index: 1;
  background: rgba(255, 255, 255, 0.95);
}

#search-bar {
  display: flex;
  flex-direction: row;
  padding: 20px 0;
  border-bottom: 1px solid var(--border-color);
}

.city-search {
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  position: relative;
  margin-right: 10px;
}

.city-search__input {
  width: 100%;
}

.city-search__suggestions {
  visibility: hidden;
  opacity: 0;
  width: 100%;
  max-height: 60vh;
  position: absolute;
  z-index: 2;
  align-self: center;
  margin: 5px 0 0;
  top: 43px;
  box-shadow: var(--heavy-shadow);
  overflow: scroll;
  background-color: var(--light-grey-no-opacity);
  border-radius: var(--border-radius);
  border: 1px solid var(--border-color);
  transition: visibility 0.15s, opacity 0.15s linear;
}

.city-search__suggestions > li {
  text-align: left;
  padding: 8px 6px 8px 10px;
  cursor: default;
  transition-duration: 0.2s;
}

.city-search__suggestions > li:last-child {
  border-bottom-left-radius: var(--border-radius);
  border-bottom-right-radius: var(--border-radius);
}

.city-search__suggestions > li:hover {
  background-color: #bcdcdb;
}

.city-search__suggestions > li:first-child:hover {
  background-color: unset;
}

.city-search__button {
  background: none;
  border: none;
  width: 100%;
  text-align: left;
  padding: 0;
  margin: 5px 0;
}

.city-search__button:hover {
  cursor: pointer;
}

.city-search__all-button {
  background: none;
  border: none;
  border-top: 1px solid var(--border-color);
  width: 100%;
  text-align: center;
  margin-top: 0;
  padding-top: 8px;
}

.city-search__all-button:hover {
  cursor: pointer;
}

.city-search__error {
  position: absolute;
  top: 50px;
  max-width: 30em;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  background-color: #fff;
  padding: 5px 10px;
  font-size: 1.25em;
  border: 1px solid var(--border-color);
  border-radius: 5px;
  box-shadow: var(--heavy-shadow);
}

.city-search__error::before {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  bottom: 100%;
  left: 1.5em;
  border: 0.75rem solid transparent;
  border-top: none;
  border-bottom-color: #fff;
  -webkit-filter: drop-shadow(0 -0.0625rem 0.0625rem rgba(0, 0, 0, 0.1));
  filter: drop-shadow(0 -0.0625rem 0.0625rem rgba(0, 0, 0, 0.1));
}

.number-of-events {
  display: flex;
  justify-content: flex-end;
  align-content: flex-start;
  font-size: var(--font-size-smaller);
  width: 50px;
  height: 38px;
}

.number-of-events__input {
  text-align: center;
  border: none;
  background: none;
  cursor: pointer;
  padding: 0;
}

#content {
  margin-top: 69px;
}

.event-list {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 0;
}

.event-list__item {
  width: 100%;
}

.event-list__item:first-of-type > div {
  margin: max(20px, 2vh) 0 2vh;
}

.event-list__warning {
  width: 100%;
  margin-top: calc(2vh - 4px);
}

.show-more-events {
  margin-top: 2vh;
  max-width: 100px;
}

.show-more-events:hover {
  cursor: pointer;
}

.show-more-events > p {
  transition: 0.3s;
}

.show-more-events:hover > p {
  transform: scale(1.1);
}

.show-more-events:hover > span {
  -webkit-animation: bounce;
  animation: bounce 0.5s infinite alternate;

  /* margin-top: 3vh; */
}

.event {
  padding: 10px;
  min-width: 250px;
  margin: 2vh 0;
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius);
  position: relative;
  box-shadow: var(--light-shadow);
  background-color: rgba(221, 246, 255, 0.1);
  transition-duration: 0.3s;
}

.event:hover,
.event:focus {
  transform: scale(1.005);
}

.event__window {
  text-align: left;
  width: 100%;
  height: 100%;
  padding: 0;
  margin: 0;
  background: none;
  border: none;
  outline: none;
  cursor: unset;
}

.event__show-more-button {
  width: 100%;
  padding: 10px 0 0;
  text-align: center;
  background: none;
  border: none;
  outline: none;
  cursor: pointer;
  transition-duration: 0.01s;
}

.turn {
  transform: rotate(-0.5turn);
}

.event__summary {
  font-size: var(--font-size-bigger);
  font-weight: 600;
}

.event__details {
  text-align: justify;
  font-weight: 300;
  background-color: var(--light-grey);
  border-radius: var(--border-radius);
  padding: 5px 10px;
}

.group-name {
  text-transform: uppercase;
}

.going,
.visibility {
  font-style: italic;
}

.event__description {
  overflow-x: hidden;
}

.button-wrapper {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 20px;
  padding: 20px;
  font-family: sans-serif;
}

.data-vis-wrapper {
  display: flex;
  align-items: center;
  flex-direction: column;
}

.data-vis-wrapper .recharts-responsive-container {
  min-width: 100%;
  display: flex;
}

.recharts-surface {
  overflow: visible;
}

.login-button {
  color: #494949;
  text-transform: uppercase;
  text-decoration: none;
  background: #fff;
  padding: 20px;
  border: var(--border-radius) solid #494949;
  display: inline-block;
  transition: all 0.4s ease 0s;
  min-width: 10rem;
  margin: 1rem;
}

.google-btn {
  width: 184px;
  height: 42px;
  margin: 1rem;
  background-color: #4285f4;
  border-radius: 2px;
  box-shadow: 0 3px 4px 0 rgba(0, 0, 0, 0.25);
}

.google-btn:hover {
  box-shadow: 0 0 6px #4285f4;
}

.google-btn:active {
  background: #1669f2;
}

.google-icon-wrapper {
  position: absolute;
  margin-top: 1px;
  margin-left: 1px;
  width: 40px;
  height: 40px;
  border-radius: 2px;
  background-color: #fff;
}

.google-icon {
  position: absolute;
  margin-top: 11px;
  margin-left: -8px;
  width: 18px;
  height: 18px;
}

.btn-text {
  float: right;
  margin: 11px 11px 0 0;
  color: #fff;
  letter-spacing: 0.2px;
  font-family: "Roboto";
}

@media only screen and (min-width: 992px) {
  .data-vis-wrapper {
    flex-direction: row;
  }

  .data-vis-wrapper .recharts-responsive-container {
    flex-basis: 50%;
    max-width: 50%;
    min-width: auto;
    margin: 0 auto;
  }
}

@-webkit-keyframes bounce {
  from {
    transform: translate3d(0, 0, 0);
  }

  to {
    transform: translate3d(0, 60%, 0);
  }
}

@keyframes bounce {
  from {
    transform: translate3d(0, 0, 0);
  }

  to {
    transform: translate3d(0, 60%, 0);
  }
}
