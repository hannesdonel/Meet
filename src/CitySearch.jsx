import React, { Component } from 'react';

class CitySearch extends Component {
  constructor() {
    super();

    this.state = {
      query: '',
      suggestions: [],
      showSuggestions: undefined,
    };
  }

  handleInputChange = (value) => {
    const { locations } = this.props;
    const suggestionsToSet = locations
      .filter((location) => location.toUpperCase().indexOf(value.toUpperCase()) > -1);
    this.setState({
      query: value,
      suggestions: suggestionsToSet,
      showSuggestions: true,
    });
  }

  handleSuggestionClick = (suggestion) => {
    const { updateEvents } = this.props;
    updateEvents(suggestion);
    this.setState({
      query: suggestion,
      showSuggestions: false,
    });
  }

  render() {
    const { query, suggestions, showSuggestions } = this.state;

    return (
      <div className="city-search">
        <input
          type="text"
          className="city-search__input"
          value={query}
          onChange={(e) => this.handleInputChange(e.target.value)}
          onFocus={() => this.setState({ showSuggestions: true })}
        />
        <ul
          className="city-search__suggestions"
          style={showSuggestions ? {} : { display: 'none' }}
        >
          {suggestions.map((suggestion) => (
            <li
              key={suggestion}
            >
              <button
                type="button"
                className="city-search__button"
                onClick={() => this.handleSuggestionClick(suggestion)}
                onKeyDown={() => this.handleSuggestionClick(suggestion)}
              >
                {suggestion}
              </button>
            </li>
          ))}
          <li key="all">
            <button
              type="button"
              className="city-search__all-button"
              onClick={() => this.handleSuggestionClick('')}
              onKeyDown={() => this.handleSuggestionClick('')}
            >
              <b>See all cities</b>

            </button>
          </li>
        </ul>
      </div>
    );
  }
}

export default CitySearch;
