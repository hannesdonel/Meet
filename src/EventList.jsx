import React from 'react';
import Event from './Event';
import { WarningAlert } from './Alert';

const EventList = ({
  events, showMore, setEventCount, count, showWarningAlert, warningText,
}) => {
  let newCount;
  const handleShowMore = () => {
    switch (count.toString()) {
      case '1':
        newCount = 2;
        break;
      case '2':
        newCount = 4;
        break;
      case '4':
        newCount = 8;
        break;
      case '8':
        newCount = 12;
        break;
      case '12':
        newCount = 16;
        break;
      case '16':
        newCount = 20;
        break;
      case '20':
        newCount = 24;
        break;
      case '24':
        newCount = 28;
        break;
      case '28':
        newCount = 32;
        break;
      case '32':
        newCount = 0;
        break;
      default:
        newCount = 0;
    }
    setEventCount(newCount);
  };

  return (
    <ul className="event-list">
      {events.map((event) => (
        <li className="event-list__item" key={event.id}>
          <Event event={event} />
        </li>
      ))}

      <div
        className={showWarningAlert ? 'event-list__warning' : 'event-list__warning display-none'}
      >
        <WarningAlert
          text={warningText}
        />
      </div>

      {showMore === true && (
        <div
          className="show-more-events"
          onClick={() => handleShowMore()}
          onKeyDown={() => handleShowMore()}
          role="button"
          tabIndex={0}
        >
          <p>show more</p>
          <span
            className="material-icons"
          >
            keyboard_double_arrow_down
          </span>
        </div>
      )}
    </ul>
  );
};

export default EventList;
