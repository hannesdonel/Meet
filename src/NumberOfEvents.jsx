import React, { Component } from 'react';

class NumberOfEvents extends Component {
  handleInputChange = (count) => {
    const { setEventCount } = this.props;
    setEventCount(count);
  }

  render() {
    return (
      <div className="number-of-events">
        <label htmlFor="number-of-events__input">
          <select
            id="number-of-events__input"
            name="number-of-events__input"
            className="number-of-events__input"
            defaultValue="32"
            onChange={(e) => this.handleInputChange(e.target.value)}
          >
            <option className="number-of-events__option" value="1">1</option>
            <option className="number-of-events__option" value="2">2</option>
            <option className="number-of-events__option" value="4">4</option>
            <option className="number-of-events__option" value="8">8</option>
            <option className="number-of-events__option" value="12">12</option>
            <option className="number-of-events__option" value="16">16</option>
            <option className="number-of-events__option" value="20">20</option>
            <option className="number-of-events__option" value="24">24</option>
            <option className="number-of-events__option" value="28">28</option>
            <option className="number-of-events__option" value="32">32</option>
            <option className="number-of-events__option" value="0">all</option>
          </select>
        </label>
      </div>
    );
  }
}

export default NumberOfEvents;
