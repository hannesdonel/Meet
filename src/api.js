export const extractLocations = (events) => {
  const extractedLocations = events.map((event) => event.location);
  const locations = [...new Set(extractedLocations)];
  return locations;
};
