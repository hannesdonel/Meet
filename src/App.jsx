import React, { Component } from 'react';
import './App.css';
import EventList from './EventList';
import CitySearch from './CitySearch';
import NumberOfEvents from './NumberOfEvents';
import WelcomeScreen from './WelcomeScreen';

import {
  extractLocations, getEvents, checkToken,
} from './api';
import './nprogress.css';
import { ErrorAlert } from './Alert';

class App extends Component {
  constructor() {
    super();

    this.state = {
      allEvents: [],
      events: [],
      locations: [],
      currentLocation: '',
      count: 32,
      isLoading: true,
      showMore: false,
      warningText: '',
      showWarningAlert: false,
      showWelcomeScreen: undefined,
    };
  }

  async componentDidMount() {
    const accessToken = localStorage.getItem('access_token');
    const { error } = await checkToken(accessToken);
    const searchParams = new URLSearchParams(window.location.search);
    const code = searchParams.get('code');
    this.setState({ showWelcomeScreen: false, isLoading: false });
    if (code || error !== 'invalid_token') {
      this.fetchData().then((data) => {
        this.setState({
          allEvents: data.events,
          events: data.events.slice(0, 32),
          locations: data.locations,
          isLoading: false,
        });
        if (data.events.length > 32) {
          this.setState({
            showMore: true,
          });
        }
      });
    }
  }

  fetchData = async () => {
    try {
      const events = await getEvents();
      const locations = extractLocations(events);
      return { events, locations };
    } catch (error) {
      return error;
    }
  };

  setEventCount = async (number) => {
    const { currentLocation } = this.state;
    await this.setState({ count: number });
    this.updateEvents(currentLocation);
  }

  setValue = () => {
    const { count } = this.state;
    const input = document.getElementById('number-of-events__input');
    input.value = count;
  }

  updateEvents = (location) => {
    const { count, allEvents } = this.state;
    const locationEvents = (location === '') ? allEvents : allEvents.filter((event) => event.location === location);

    if (count.toString() === '0') {
      this.setValue();
      this.setState({
        currentLocation: location,
        events: locationEvents,
        showMore: false,
        showWarningAlert: false,
      });
    } else if (count.toString() !== '0') {
      const locationEventsShortened = locationEvents.slice(0, count);
      this.setValue();
      if (locationEventsShortened.length === locationEvents.length) {
        this.setState({
          currentLocation: location,
          events: locationEventsShortened,
          showMore: false,
          showWarningAlert: false,
        });
      } else {
        this.setState({
          currentLocation: location,
          events: locationEventsShortened,
          showMore: true,
          showWarningAlert: false,
        });
      }
      if (locationEvents.length < parseInt(count, 10)) {
        this.setState({
          warningText: 'Your request did not produce any more results.',
          showWarningAlert: true,
        });
      }
    }
  }

  toggleLoadingScreen = (boolean) => {
    this.setState({
      isLoading: boolean,
    });
  }

  render() {
    const {
      events, locations, showMore, count, updateEvents,
      isLoading, showWarningAlert, warningText, showWelcomeScreen,
    } = this.state;

    if (isLoading) {
      return (
        <div className="loader-container"><p>loading...</p></div>
      );
    }

    if (showWelcomeScreen) {
      return (
        <WelcomeScreen
          showWelcomeScreen={showWelcomeScreen}
          isLoading={this.toggleLoadingScreen}
        />
      );
    }

    return (
      <div className="App">
        <div>
          <div id="search-bar-container">
            <div id="search-bar">
              <CitySearch locations={locations} updateEvents={updateEvents} />
              <NumberOfEvents setEventCount={this.setEventCount} />
            </div>
          </div>
          <div id="content">
            <div
              className={!navigator.onLine ? 'offline-alert' : 'offline-alert display-none'}
            >
              <ErrorAlert text="It seems you're offline." color="#ffffff" />
            </div>
            <EventList
              events={events}
              showMore={showMore}
              setEventCount={this.setEventCount}
              count={count}
              showWarningAlert={showWarningAlert}
              warningText={warningText}
            />
          </div>
        </div>
      </div>
    );
  }
}

export default App;
