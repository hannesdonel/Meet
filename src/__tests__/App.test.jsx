import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import App from '../App';
import EventList from '../EventList';
import CitySearch from '../CitySearch';

import mockData from '../mock-data';
import { extractLocations, getEvents } from '../api';
import NumberOfEvents from '../NumberOfEvents';

describe('<App /> component', () => {
  let AppWrapper;
  beforeAll(() => {
    AppWrapper = shallow(<App />);
  });

  afterAll(() => {
    AppWrapper.unmount();
  });

  test('render App correctly', () => {
    const tree = renderer.create(<App />).toJSON;
    expect(tree).toMatchSnapshot();
  });

  test('setEventCount sets "count" state correctly', () => {
    const number = 20;
    AppWrapper.instance().setEventCount(number);
    expect(AppWrapper.state('count')).toEqual(number);
  });
});

describe('<App /> integration', () => {
  test('App passes "events" state as a prop to EventList', () => {
    const AppWrapper = mount(<App />);
    const AppEventsState = AppWrapper.state('events');
    expect(AppEventsState).not.toEqual(undefined);
    expect(AppWrapper.find(EventList).props().events).toEqual(AppEventsState);
    AppWrapper.unmount();
  });

  test('App passes "locations" state as a prop to CitySearch', () => {
    const AppWrapper = mount(<App />);
    const AppLocationsState = AppWrapper.state('locations');
    expect(AppLocationsState).not.toEqual(undefined);
    expect(AppWrapper.find(CitySearch).props().locations).toEqual(AppLocationsState);
    AppWrapper.unmount();
  });

  test('get list of events matching the city selected by the user', async () => {
    const AppWrapper = mount(<App />);
    const CitySearchWrapper = AppWrapper.find(CitySearch);
    const locations = extractLocations(mockData);
    CitySearchWrapper.setState({ suggestions: locations });
    const suggestions = CitySearchWrapper.state('suggestions');
    const selectedIndex = Math.floor(Math.random() * (suggestions.length));
    const selectedCity = suggestions[selectedIndex];
    await CitySearchWrapper.instance().handleSuggestionClick(selectedCity);
    const allEvents = await getEvents();
    const eventsToShow = allEvents.filter((event) => event.location === selectedCity);
    expect(AppWrapper.state('events')).toEqual(eventsToShow);
    AppWrapper.unmount();
  });

  test('get list of all events when user selects "See all cities"', async () => {
    const AppWrapper = mount(<App />);
    const suggestionItems = AppWrapper.find(CitySearch).find('.city-search__suggestions button');
    await suggestionItems.at(suggestionItems.length - 1).simulate('click');
    const allEvents = await getEvents();
    expect(AppWrapper.state('events')).toEqual(allEvents);
    AppWrapper.unmount();
  });

  test('NumberOfEvents sets number of events to be displayed correctly', async () => {
    const AppWrapper = mount(<App />);
    const NumberOfEventesWrapper = AppWrapper.find(NumberOfEvents);
    await NumberOfEventesWrapper.find('.number-of-events__input').simulate('change', { target: { value: 1 } });
    const countState = AppWrapper.state('count');
    const eventsState = AppWrapper.state('events');
    expect(eventsState.length).toEqual(countState);
    AppWrapper.unmount();
  });

  test('show all events when NumberOfEvents is set to "0"', async () => {
    const AppWrapper = await mount(<App />);
    const NumberOfEventesWrapper = AppWrapper.find(NumberOfEvents);
    const eventsStateBefore = AppWrapper.state('events');
    await NumberOfEventesWrapper.find('.number-of-events__input').simulate('change', { target: { value: '0' } });
    const eventsStateAfter = AppWrapper.state('events');
    expect(JSON.stringify(eventsStateAfter)).toMatch(JSON.stringify(eventsStateBefore));
    AppWrapper.unmount();
  });
});
