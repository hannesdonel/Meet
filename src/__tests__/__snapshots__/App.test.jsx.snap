// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> component render App correctly 1`] = `[Function]`;
